### PR TITLE
[stable/oauth2-proxy] More configurations and fixes

### PR DESCRIPTION
**Pull request by victornoel**
#### What this PR does / why we need it:

On top of fixing some documentation and naming, this PR adds:
- the possibility to configure a Google service account
- to specify a `ConfigMap` for oauth2-proxy configuration (instead of inlining it in the values)
- move default configuration from `extraArgs` to the default `configFile`
  - it is unusable as-is if one wants to set the upstream or the domains via the config file because the arguments take over the configuration
  - :warning:  this maybe is breaking change :warning: 
- add checksums of the configuration in annotations for automatic reload of the pod when needed

#### Special notes for your reviewer:

@Miouge1 tell me how the chart version number should be bumped :)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [ ] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by victornoel** [comment]
@Miouge1 @unguiculus hi, anyway to get some feedback on this? :)

File: stable/oauth2-proxy/values.yaml [hunk 1]
```diff
@@ -5,16 +5,28 @@ config:
...
+    upstreams = [ "file:///dev/null" ]
   # Custom configuration file: oauth2_proxy.cfg
   # configFile: |-
   #   pass_basic_auth = false
   #   pass_access_token = true
-  configFile: ""
-  # Use an existing secret for OAuth2 credentials
+  # Use an existing config map (see configmap.yaml for required fields)
   # Example:
-  # existingSecret: secret
```
Comment (Miouge1) [inline stable/oauth2-proxy/values.yaml:17]: `config.existingSecret` is still a valid thing no?

File: stable/oauth2-proxy/templates/deployment.yaml [hunk 2]
```diff
@@ -51,6 +55,12 @@ spec:
           - --authenticated-emails-file=/etc/oauth2-proxy/authenticated-emails-list
         {{- end }}
         {{- end }}
+        {{- with .Values.config.google }}
+        {{- if and .adminEmail (or .serviceAccountJson .existingSecret) }}
+          - --google-admin-email={{ .adminEmail }}
+          - --google-service-account-json=/google/service-account.json
```
Comment (Miouge1) [inline stable/oauth2-proxy/templates/deployment.yaml:62]: I can't see where is this `/google/` path mounted?

**Review by Miouge1 — COMMENTED** [review]
@victornoel thank you for your contribution and sorry for the delay: I was absent last week.

You'll need to bump the chart version, since this adds a new feature, I think it will be `X.(Y+1).Z`

File: stable/oauth2-proxy/values.yaml (hunk 1 above)
Comment (victornoel) [reply stable/oauth2-proxy/values.yaml:17]: It was moved a few line above actually, the diff doesn't show it well I believe but it still there :)

File: stable/oauth2-proxy/templates/deployment.yaml (hunk 2 above)
Comment (victornoel) [reply stable/oauth2-proxy/templates/deployment.yaml:62]: A few line below under the `volumeMounts` property

**Comment by victornoel** [comment]
@Miouge1 see my answers to your comment above. I've also added a commit to bump the version to `0.10.0`.

**Comment by Miouge1** [comment]
Another PR already bumbed the version to `0.10.0`, so you need `0.11.0`

**Comment by victornoel** [comment]
@Miouge1 here it is (and rebased onto master)

File: stable/oauth2-proxy/values.yaml [hunk 3]
```diff
@@ -28,8 +40,6 @@ image:
   # - name: myRegistryKeySecretName
 
 extraArgs:
```
Comment (Miouge1) [inline stable/oauth2-proxy/values.yaml]: Accoridng to the doc the default value is `{}`

**Comment by Miouge1** [comment]
I did a test upgrade from current version to this patch, it worked fine. Before merging, I'm considering if this is backwards compatible or not, I see 4 situations:
| extraArgs | configFile | Outcome |
|-----------|-----------|----------|
| Default | Default | ✅  |
| Custom | Default | ✅  |
| Default | Custom | 🔴   |
| Custom | Custom | 🔴   |

Is that correct?

File: stable/oauth2-proxy/values.yaml (hunk 3 above)
Comment (victornoel) [reply stable/oauth2-proxy/values.yaml]: @Miouge1 you are right, so the documentation was incorrect before, what do you think we should do? Prefer to set it to `{}` or put the default args in the documentation?

I feel like the current value of `extraArgs` is a value we want to have even if the user override `extraArgs`, so I wonder if we shouldn't instead put this argument directly in the deployment and simply have `extraArgs` default to `{}`. WDYT?

**Comment by victornoel** [comment]
> I did a test upgrade from current version to this patch, it worked fine. Before merging, I'm considering if this is backwards compatible or not, I see 4 situations:
> extraArgs 	configFile 	backwards compatible?
> Default 	Default 	white_check_mark
> Custom 	Default 	white_check_mark
> Default 	Custom 	red_circle
> Custom 	Custom 	red_circle
> 
> Is that correct?

I think for the second one, it's correct as long as we expect the args to take over the configuration (which is what I have observed).

For the third one, it depends on what was in the config file, but mostly yes, it's not backward compatible. The last one is similar to this one.

So I agree with this table :)

File: stable/oauth2-proxy/values.yaml (hunk 3 above)
Comment (Miouge1) [reply stable/oauth2-proxy/values.yaml]: Considering that the `containerPort: 4180` is fixed in `templates/deployment.yaml`, I agree that it makes sense to set the `http-address` on top of the `extraArgs`

**Comment by Miouge1** [comment]
OK, then as per [SemVer](https://semver.org/), I suggest to release this as version `1.0.0`.

**Comment by victornoel** [comment]
> OK, then as per [SemVer](https://semver.org/), I suggest to release this as version `1.0.0`.

An alternative is to consider that versions starting with `0.` are non stable and thus can be breaking when you change the second digit: see https://semver.org/#spec-item-4.

In that case, it would be better to keep this in `0.x.x` until we are sure we are not going to break more things :)

File: stable/oauth2-proxy/values.yaml (hunk 3 above)
Comment (victornoel) [reply stable/oauth2-proxy/values.yaml]: @Miouge1 it's pushed :)

**Comment by Miouge1** [comment]
Either way works for me.

**Comment by victornoel** [comment]
@Miouge1 then let's keep it at 0.11.0 as it is now and merge it if there are no other changes to make :)

**Comment by Miouge1** [comment]
Sounds good.

/approved
/lgtm

**Pull request merged**